### PR TITLE
ActionSchedule & Conditional field validation

### PR DESCRIPTION
**Pull request by monishdeb**

### CODE REVIEW

File: Civi/Api4/Action/Create.php [hunk 1]
```diff
@@ -40,6 +40,13 @@
  */
 class Create extends AbstractAction {
 
+  /**
+   * And one of the field in the set is required
+   *
+   * @var array
+   */
+  protected $optionalRequiredFields = [];
```
Comment (monishdeb) [inline Civi/Api4/Action/Create.php]: @colemanw this special variable to support ```civicrm_api3_verify_one_mandatory``` where either one the attribute is mandatory. Do you agree with this is the approach?

Comment (colemanw) [reply Civi/Api4/Action/Create.php]: I think adding it as a protected variable makes it an api param and exposes it to the api explorer. We don't want that.

I would prefer a metadata-driven approach. @monishdeb @eileenmcnaughton what would you think about adding a php-like syntax to a specal `requiredIf` fieldspec property, like this:

    $fields = array(
      'contact_type' => [
        'title' => 'Contact Type',
        'data_type' => 'String',
        'required' => true,
      ],
      'first_name' => [
        'title' => 'First Name',
        'data_type' => 'String',
        'requiredIf' => "$contact_type == 'Individual' && !$last_name && !$email",
      ],
      'last_name' => [
        'title' => 'Last Name',
        'data_type' => 'String',
        'requiredIf' => "$contact_type == 'Individual' && !$first_name && !$email",
      ],
      'organization_name' => [
        'title' => 'Organization Name',
        'data_type' => 'String',
        'requiredIf' => "$contact_type == 'Organization' && !$email",
      ],
      'household_name' => [
        'title' => 'Household Name',
        'data_type' => 'String',
        'requiredIf' => "$contact_type == 'Household' && !$email",
      ],
    );

Comment (colemanw) [reply Civi/Api4/Action/Create.php]: We might even be able to use the Smarty engine to parse those conditional strings.

Comment (monishdeb) [reply Civi/Api4/Action/Create.php]: I like this approach, and if I am not wrong it should extend the ``validateValues``` and use ```eval()``` to translate the expression.

Comment (colemanw) [reply Civi/Api4/Action/Create.php]: I think the evaluator code could look something like this:

    function evaluateExpression($expr, $vars) {
      var $smarty = new Smarty();
      foreach ($vars as $key => $val) {
        $smarty->assign($key, $val);
      }
      $tpl = "{if $expr}1{else}0{/if}";
      $result = trim($smarty->fetch('eval:' . $tpl));
      return $result;
    }

**Comment by monishdeb** [comment]
@colemanw sorry for the delay. This is my last [commit](https://github.com/civicrm/org.civicrm.api4/pull/92/commits/0d51325930b4dba9efb50121b0584f962eb2dd40) which implements conditional requiredness of a attribute by evaluating the smary exp. Here's a screencast how does it behave for Contact.create API on choosing desired contact types, similarly for ActionSchedule.create:
before

**Comment by seamuslee001** [comment]
@monishdeb there is a test failure here 

```
                                                                               
    thrown in /home/jenkins/bknix-dfl/build/ext-3/sites/default/files/civicrm  
  /ext/target/Civi/Api4/Action/Create.php on line 92                           
                                                                               

Exception trace:
 Civici\Util\Process::runOk() at /opt/civici/src/Util/ProcessBatch.php:65
 Civici\Util\ProcessBatch->runAllOk() at /opt/civici/src/Command/ExtTestCommand.php:125
 Civici\Command\ExtTestCommand->execute() at /opt/civici/vendor/symfony/console/Command/Command.php:251
 Symfony\Component\Console\Command\Command->run() at /opt/civici/vendor/symfony/console/Application.php:946
 Symfony\Component\Console\Application->doRunCommand() at /opt/civici/vendor/symfony/console/Application.php:248
 Symfony\Component\Console\Application->doRun() at /opt/civici/vendor/symfony/console/Application.php:148
 Symfony\Component\Console\Application->run() at /opt/civici/src/Application.php:17
 Civici\Application::main() at /opt/civici/bin/civici:19

ext:test [-N|--dry-run] [--info INFO] [--junit-dir JUNIT-DIR] [--timeout TIMEOUT]
```

**Comment by seamuslee001** [comment]
@totten any reason why even tho jenkins set it to failure there is a check here?

**Comment by colemanw** [comment]
@monishdeb this was getting hung up on the conditional requirements so I've pushed the other entities separately and rebased this to only include ActionSchedule which is the one that needs the conditionals.

Also FYI I changed `CRM_Core_Smarty::singleton()` to `new CRM_Core_Smarty()` because we don't want to create weird bugs when reusing templates that already have variables assigned to them.

I think the next step is to add some unit tests around the `validateValues()` function to make sure it is correctly evaluating the conditional expressions. Can you work on that?

**Comment by colemanw** [comment]
I've added a test.

**Pull request merged**